<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width,initial-scale=1">
    <title>Rcwl/RcwlPipelines case study: scRNA-seq data preprocessing | Rcwl</title>
    <meta name="generator" content="VuePress 1.9.7">
    
    <meta name="description" content="R and Bioconductor toolchain for bioinformatics tools and workflows">
    <meta name="theme-color" content="#3eaf7c">
    <meta name="apple-mobile-web-app-capable" content="yes">
    <meta name="apple-mobile-web-app-status-bar-style" content="black">
    
    <link rel="preload" href="/assets/css/0.styles.2e55c9e9.css" as="style"><link rel="preload" href="/assets/js/app.78d66748.js" as="script"><link rel="preload" href="/assets/js/2.77543436.js" as="script"><link rel="preload" href="/assets/js/61.31d1b0d6.js" as="script"><link rel="prefetch" href="/assets/js/10.687810b6.js"><link rel="prefetch" href="/assets/js/11.5949a480.js"><link rel="prefetch" href="/assets/js/12.3dcf05a7.js"><link rel="prefetch" href="/assets/js/13.7591d52c.js"><link rel="prefetch" href="/assets/js/14.b61a8e66.js"><link rel="prefetch" href="/assets/js/15.7b9230a9.js"><link rel="prefetch" href="/assets/js/16.df975473.js"><link rel="prefetch" href="/assets/js/17.25fea857.js"><link rel="prefetch" href="/assets/js/18.a4764f16.js"><link rel="prefetch" href="/assets/js/19.7c746131.js"><link rel="prefetch" href="/assets/js/20.028a35d4.js"><link rel="prefetch" href="/assets/js/21.2b63d498.js"><link rel="prefetch" href="/assets/js/22.5ebd8b55.js"><link rel="prefetch" href="/assets/js/23.7ce61c7b.js"><link rel="prefetch" href="/assets/js/24.9e44104a.js"><link rel="prefetch" href="/assets/js/25.8e228e57.js"><link rel="prefetch" href="/assets/js/26.497d563f.js"><link rel="prefetch" href="/assets/js/27.b013b216.js"><link rel="prefetch" href="/assets/js/28.a3b0c631.js"><link rel="prefetch" href="/assets/js/29.38024892.js"><link rel="prefetch" href="/assets/js/3.10c9a18c.js"><link rel="prefetch" href="/assets/js/30.af9f0acf.js"><link rel="prefetch" href="/assets/js/31.6ecb86cf.js"><link rel="prefetch" href="/assets/js/32.0ffea07c.js"><link rel="prefetch" href="/assets/js/33.a19168e4.js"><link rel="prefetch" href="/assets/js/34.92cfd236.js"><link rel="prefetch" href="/assets/js/35.f3902218.js"><link rel="prefetch" href="/assets/js/36.dc547c77.js"><link rel="prefetch" href="/assets/js/37.4f74995e.js"><link rel="prefetch" href="/assets/js/38.be790fff.js"><link rel="prefetch" href="/assets/js/39.8cbd959b.js"><link rel="prefetch" href="/assets/js/4.b4ea30d7.js"><link rel="prefetch" href="/assets/js/40.06538e3a.js"><link rel="prefetch" href="/assets/js/41.3993e7fa.js"><link rel="prefetch" href="/assets/js/42.ccbe0e60.js"><link rel="prefetch" href="/assets/js/43.66d4cc36.js"><link rel="prefetch" href="/assets/js/44.ea84bd3d.js"><link rel="prefetch" href="/assets/js/45.0575bedd.js"><link rel="prefetch" href="/assets/js/46.33ebacc8.js"><link rel="prefetch" href="/assets/js/47.ff128a87.js"><link rel="prefetch" href="/assets/js/48.fef3ccac.js"><link rel="prefetch" href="/assets/js/49.c525b7b2.js"><link rel="prefetch" href="/assets/js/5.9c6b78e0.js"><link rel="prefetch" href="/assets/js/50.c0518109.js"><link rel="prefetch" href="/assets/js/51.1366bb01.js"><link rel="prefetch" href="/assets/js/52.5a116fe8.js"><link rel="prefetch" href="/assets/js/53.7a598e47.js"><link rel="prefetch" href="/assets/js/54.74928571.js"><link rel="prefetch" href="/assets/js/55.bc202844.js"><link rel="prefetch" href="/assets/js/56.a1e667d9.js"><link rel="prefetch" href="/assets/js/57.aa9cf87c.js"><link rel="prefetch" href="/assets/js/58.ec09dea8.js"><link rel="prefetch" href="/assets/js/59.14ee9c1a.js"><link rel="prefetch" href="/assets/js/6.cd18b48d.js"><link rel="prefetch" href="/assets/js/60.5d8cfb7b.js"><link rel="prefetch" href="/assets/js/62.71ea3aca.js"><link rel="prefetch" href="/assets/js/63.acf63e83.js"><link rel="prefetch" href="/assets/js/64.c3965751.js"><link rel="prefetch" href="/assets/js/65.8cc1ef07.js"><link rel="prefetch" href="/assets/js/66.9f1574bd.js"><link rel="prefetch" href="/assets/js/67.c458a52d.js"><link rel="prefetch" href="/assets/js/68.70f932c2.js"><link rel="prefetch" href="/assets/js/69.3346a70a.js"><link rel="prefetch" href="/assets/js/7.9a0cf5d9.js"><link rel="prefetch" href="/assets/js/70.5579bade.js"><link rel="prefetch" href="/assets/js/71.31927857.js"><link rel="prefetch" href="/assets/js/72.17df1f49.js"><link rel="prefetch" href="/assets/js/73.86c609d0.js"><link rel="prefetch" href="/assets/js/74.4be915a5.js"><link rel="prefetch" href="/assets/js/75.d3bc8588.js"><link rel="prefetch" href="/assets/js/76.438110ff.js"><link rel="prefetch" href="/assets/js/77.7139f333.js"><link rel="prefetch" href="/assets/js/78.96037bb8.js"><link rel="prefetch" href="/assets/js/79.51d67b58.js"><link rel="prefetch" href="/assets/js/8.be3f9fd9.js"><link rel="prefetch" href="/assets/js/80.eee4f971.js"><link rel="prefetch" href="/assets/js/81.58e93231.js"><link rel="prefetch" href="/assets/js/82.710f9413.js"><link rel="prefetch" href="/assets/js/9.8aa512a8.js">
    <link rel="stylesheet" href="/assets/css/0.styles.2e55c9e9.css">
  </head>
  <body>
    <div id="app" data-server-rendered="true"><div class="theme-container"><header class="navbar"><div class="sidebar-button"><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" role="img" viewBox="0 0 448 512" class="icon"><path fill="currentColor" d="M436 124H12c-6.627 0-12-5.373-12-12V80c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12zm0 160H12c-6.627 0-12-5.373-12-12v-32c0-6.627 5.373-12 12-12h424c6.627 0 12 5.373 12 12v32c0 6.627-5.373 12-12 12z"></path></svg></div> <a href="/" class="home-link router-link-active"><!----> <span class="site-name">Rcwl</span></a> <div class="links"><div class="search-box"><input aria-label="Search" autocomplete="off" spellcheck="false" value=""> <!----></div> <nav class="nav-links can-hide"><div class="nav-item"><a href="/guide/" class="nav-link">
  Guide
</a></div><div class="nav-item"><a href="/RcwlRecipes/" class="nav-link">
  Rcwl Pipelines
</a></div><div class="nav-item"><a href="/dataRecipes/" class="nav-link">
  Data Recipes
</a></div><div class="nav-item"><div class="dropdown-wrapper"><button type="button" aria-label="Case studies" class="dropdown-title"><span class="title">Case studies</span> <span class="arrow down"></span></button> <button type="button" aria-label="Case studies" class="mobile-dropdown-title"><span class="title">Case studies</span> <span class="arrow right"></span></button> <ul class="nav-dropdown" style="display:none;"><li class="dropdown-item"><!----> <a href="/case_studies/scrnaseq/" aria-current="page" class="nav-link router-link-exact-active router-link-active">
  scRNA-seq preprocessing
</a></li></ul></div></div><div class="nav-item"><a href="https://rworkflow.github.io/RcwlBook/" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Tutorials
  <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></div><div class="nav-item"><a href="https://rworkflow.github.io/Rcwl/" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Rcwl
  <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></div><div class="nav-item"><a href="https://rworkflow.github.io/RcwlPipelines/" target="_blank" rel="noopener noreferrer" class="nav-link external">
  RcwlPipelines
  <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></div><div class="nav-item"><a href="/about/" class="nav-link">
  Contact Us
</a></div> <!----></nav></div></header> <div class="sidebar-mask"></div> <aside class="sidebar"><nav class="nav-links"><div class="nav-item"><a href="/guide/" class="nav-link">
  Guide
</a></div><div class="nav-item"><a href="/RcwlRecipes/" class="nav-link">
  Rcwl Pipelines
</a></div><div class="nav-item"><a href="/dataRecipes/" class="nav-link">
  Data Recipes
</a></div><div class="nav-item"><div class="dropdown-wrapper"><button type="button" aria-label="Case studies" class="dropdown-title"><span class="title">Case studies</span> <span class="arrow down"></span></button> <button type="button" aria-label="Case studies" class="mobile-dropdown-title"><span class="title">Case studies</span> <span class="arrow right"></span></button> <ul class="nav-dropdown" style="display:none;"><li class="dropdown-item"><!----> <a href="/case_studies/scrnaseq/" aria-current="page" class="nav-link router-link-exact-active router-link-active">
  scRNA-seq preprocessing
</a></li></ul></div></div><div class="nav-item"><a href="https://rworkflow.github.io/RcwlBook/" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Tutorials
  <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></div><div class="nav-item"><a href="https://rworkflow.github.io/Rcwl/" target="_blank" rel="noopener noreferrer" class="nav-link external">
  Rcwl
  <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></div><div class="nav-item"><a href="https://rworkflow.github.io/RcwlPipelines/" target="_blank" rel="noopener noreferrer" class="nav-link external">
  RcwlPipelines
  <span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a></div><div class="nav-item"><a href="/about/" class="nav-link">
  Contact Us
</a></div> <!----></nav>  <ul class="sidebar-links"><li><section class="sidebar-group depth-0"><p class="sidebar-heading open"><span>Scrnaseq</span> <!----></p> <ul class="sidebar-links sidebar-group-items"><li><a href="/case_studies/scrnaseq/" aria-current="page" class="active sidebar-link">Rcwl/RcwlPipelines case study: scRNA-seq data preprocessing</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#introduction" class="sidebar-link">Introduction</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#scrna-seq-data-source" class="sidebar-link">scRNA-seq data source</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#load-the-packages-and-sync-all-tools" class="sidebar-link">Load the packages and sync all tools</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#install-packages" class="sidebar-link">Install packages</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#load-scrna-seq-pre-processing-tools" class="sidebar-link">Load scRNA-seq pre-processing tools</a></li></ul></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#scrna-seq-preprocessing" class="sidebar-link">scRNA-seq preprocessing</a><ul class="sidebar-sub-headers"><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#indexing" class="sidebar-link">Indexing</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#alignment" class="sidebar-link">Alignment</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#count-filtering" class="sidebar-link">Count filtering</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#scrna-seq-preprocessing-pipeline" class="sidebar-link">scRNA-seq preprocessing pipeline</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#submit-parallel-jobs-for-starsolo" class="sidebar-link">Submit parallel jobs for STARsolo</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#shiny-interface" class="sidebar-link">Shiny interface</a></li></ul></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#sessioninfo" class="sidebar-link">SessionInfo</a></li><li class="sidebar-sub-header"><a href="/case_studies/scrnaseq/#links" class="sidebar-link">Links</a></li></ul></li></ul></section></li></ul> </aside> <main class="page"> <div class="theme-default-content content__default"><h1 id="rcwl-rcwlpipelines-case-study-scrna-seq-data-preprocessing"><a href="#rcwl-rcwlpipelines-case-study-scrna-seq-data-preprocessing" class="header-anchor">#</a> Rcwl/RcwlPipelines case study: scRNA-seq data preprocessing</h1> <h2 id="introduction"><a href="#introduction" class="header-anchor">#</a> Introduction</h2> <p>Here we demonstrate a case study for scRNA-seq data preprocessing
using <code>RcwlPipeline</code> tools and pipeline.</p> <p>10x Genomics has its own preprocessing pipeline <code>Cell Ranger</code> to
process the scRNA-seq outputs it produces to perform the
demultiplexing and quantification. However, it requires much
configuration to run and is significantly slower than other mappers.</p> <p>In this case study,
<a href="https://github.com/alexdobin/STAR/blob/master/docs/STARsolo.md" target="_blank" rel="noopener noreferrer">STARsolo<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>
is used for alignment and quantification, which produces a count
matrix from FASTQ.
<a href="https://bioconductor.org/packages/DropletUtils/" target="_blank" rel="noopener noreferrer">DropletUtils<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> is
used for filtering raw gene-barcode matrix and removing empty
droplets, which produces a high-quality count matrix with feature/cell
annotation files saved in an <em>R</em> object of <code>SingleCellExperiment</code>.
Before these, a one-time indexing step using
<a href="https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/03_alignment.html" target="_blank" rel="noopener noreferrer">STARindex<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>
is also included in this case study.</p> <h2 id="scrna-seq-data-source"><a href="#scrna-seq-data-source" class="header-anchor">#</a> scRNA-seq data source</h2> <p>The scRNA-seq data source is the 1k PBMCs from 10x genomics (These
source files are provided in the
<a href="https://zenodo.org/record/3457880" target="_blank" rel="noopener noreferrer">Zenodo<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> data repository).</p> <p>The dataset used in this tutorial are sub-sampled from the source
files to contain only 15 cells instead of 1000. The data curation is
for demo purpose only so that the execution of the Rcwl scRNA-seq
preprocessing tools or pipeline in R can be completed within 1~2
minutes.</p> <p>The &quot;*.fastq&quot; data was curated to only include reads on chromosome 21.</p> <p>“subset15_demo_barcode.txt” contains known cell barcodes for mapping
and only 15 barcodes are included.</p> <p>&quot;Homo_sapiens.GRCh37.75.21.gtf&quot; contains the hg19 GTF file to annotate
reads, which was curated on chromosome 21 only.</p> <p>Data can be loaded from the dedicated <a href="https://github.com/rworkflow/testdata" target="_blank" rel="noopener noreferrer">GitHub
repository<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> or the
<a href="https://zenodo.org/record/4573881" target="_blank" rel="noopener noreferrer">Zenodo<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> data repository.</p> <div class="language-r extra-class"><pre class="language-r"><code>library<span class="token punctuation">(</span>git2r<span class="token punctuation">)</span>
clone<span class="token punctuation">(</span><span class="token string">&quot;https://github.com/rworkflow/testdata&quot;</span><span class="token punctuation">,</span> <span class="token string">&quot;rcwl_data_supp&quot;</span><span class="token punctuation">)</span>
</code></pre></div><div class="language-r extra-class"><pre class="language-r"><code>path <span class="token operator">&lt;-</span> <span class="token string">&quot;rcwl_data_supp&quot;</span>  <span class="token comment">## source data</span>
dir<span class="token punctuation">(</span>path<span class="token punctuation">)</span>
</code></pre></div><p>Here we also create an output directory to save result files from
running the tool/pipeline.</p> <div class="language-r extra-class"><pre class="language-r"><code>outpath <span class="token operator">&lt;-</span> <span class="token string">&quot;outdir&quot;</span> 
dir.create<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> showWarnings <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">)</span>
</code></pre></div><h2 id="load-the-packages-and-sync-all-tools"><a href="#load-the-packages-and-sync-all-tools" class="header-anchor">#</a> Load the packages and sync all tools</h2> <h3 id="install-packages"><a href="#install-packages" class="header-anchor">#</a> Install packages</h3> <p>If this is the first time you use <code>Rcwl</code> or <code>RcwlPipelines</code>, you need
to install the packages:</p> <p>The package can be installed from Bioconductor (&gt;= 3.9):</p> <div class="language-r extra-class"><pre class="language-r"><code><span class="token keyword">if</span> <span class="token punctuation">(</span><span class="token operator">!</span>requireNamespace<span class="token punctuation">(</span><span class="token string">&quot;BiocManager&quot;</span><span class="token punctuation">,</span> quietly <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
    install.packages<span class="token punctuation">(</span><span class="token string">&quot;BiocManager&quot;</span><span class="token punctuation">)</span>
BiocManager<span class="token operator">::</span>install<span class="token punctuation">(</span>c<span class="token punctuation">(</span><span class="token string">&quot;Rcwl&quot;</span><span class="token punctuation">,</span> <span class="token string">&quot;RcwlPipelines&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre></div><p>Or use the most updated version from GitHub:</p> <div class="language-r extra-class"><pre class="language-r"><code>BiocManager<span class="token operator">::</span>install<span class="token punctuation">(</span>c<span class="token punctuation">(</span><span class="token string">&quot;rworkflow/Rcwl&quot;</span><span class="token punctuation">,</span> <span class="token string">&quot;rworkflow/RcwlPipelines&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre></div><p>Load the packages:</p> <div class="language-r extra-class"><pre class="language-r"><code>library<span class="token punctuation">(</span>Rcwl<span class="token punctuation">)</span>
library<span class="token punctuation">(</span>RcwlPipelines<span class="token punctuation">)</span>
</code></pre></div><h3 id="load-scrna-seq-pre-processing-tools"><a href="#load-scrna-seq-pre-processing-tools" class="header-anchor">#</a> Load scRNA-seq pre-processing tools</h3> <p>Three core functions: <code>cwlUpdate</code>, <code>cwlSearch</code> and <code>cwlLoad</code>from
<code>RcwlPipelines</code> will be needed for updating, searching, and loading
the needed tools or pipelines in <em>R</em>.</p> <p>The <code>cwlUpdate</code> function syncs the current <code>Rcwl</code> recipes and returns
a <code>cwlHub</code> object which contains the most updated <code>Rcwl</code> recipes. The
<code>mcols()</code> function returns all related information about each
available tool or pipeline.  The recipes will be locally cached, so
users don't need to call <code>cwlUpdate</code> every time unless they want to
use a tool/pipeline that is newly added to <code>RcwlPipelines</code>.</p> <p>In this example, we are using the Bioc 3.13 which is the current devel
version.</p> <div class="language-r extra-class"><pre class="language-r"><code>atls <span class="token operator">&lt;-</span> cwlUpdate<span class="token punctuation">(</span>branch <span class="token operator">=</span> <span class="token string">&quot;dev&quot;</span><span class="token punctuation">,</span> force <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span> <span class="token comment">## sync the tools/pipelines. </span>
atls
table<span class="token punctuation">(</span>mcols<span class="token punctuation">(</span>atls<span class="token punctuation">)</span><span class="token operator">$</span>Type<span class="token punctuation">)</span>
</code></pre></div><p><code>cwlSearch</code> is used to search for specific tools/pipelines of
interest. Multiple keywords can be used for an internal search of
&quot;rname&quot;, &quot;rpath&quot;, &quot;fpath&quot;, &quot;Command&quot; and &quot;Containers&quot; columns in the
<code>mcols()</code>.</p> <div class="language-r extra-class"><pre class="language-r"><code>tls <span class="token operator">&lt;-</span> cwlSearch<span class="token punctuation">(</span>c<span class="token punctuation">(</span><span class="token string">&quot;STAR&quot;</span><span class="token punctuation">,</span> <span class="token string">&quot;index&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
mcols<span class="token punctuation">(</span>tls<span class="token punctuation">)</span>
</code></pre></div><p><code>cwlLoad</code> loads the <code>Rcwl</code> tool/pipeline into the <em>R</em> working
environment. Here we load all the 3 tools that will be needed for the
scRNA-seq data preprocessing. The recipes for developing these tools
can be found in the <a href="https://github.com/rworkflow/RcwlRecipes/tree/master/Rcwl" target="_blank" rel="noopener noreferrer">GitHub
repository<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>
including
<a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/tl_STARindex.R" target="_blank" rel="noopener noreferrer">tl_STARindex<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>,
<a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/tl_STARsolo.R" target="_blank" rel="noopener noreferrer">tl_STARsolo<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>,
and
<a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/tl_DropletUtils.R" target="_blank" rel="noopener noreferrer">tl_DropletUtils<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</p> <div class="language-r extra-class"><pre class="language-r"><code>STARindex <span class="token operator">&lt;-</span> cwlLoad<span class="token punctuation">(</span>title<span class="token punctuation">(</span>tls<span class="token punctuation">)</span><span class="token punctuation">[</span><span class="token number">2</span><span class="token punctuation">]</span><span class="token punctuation">)</span>  <span class="token comment">## &quot;tl_STARindex&quot;</span>
<span class="token comment">## equivalent:</span>
<span class="token comment">## STARindex &lt;- cwlLoad(mcols(tls)$fpath[2])</span>
STARsolo <span class="token operator">&lt;-</span> cwlLoad<span class="token punctuation">(</span><span class="token string">&quot;tl_STARsolo&quot;</span><span class="token punctuation">)</span>
DropletUtils <span class="token operator">&lt;-</span> cwlLoad<span class="token punctuation">(</span><span class="token string">&quot;tl_DropletUtils&quot;</span><span class="token punctuation">)</span>
</code></pre></div><h2 id="scrna-seq-preprocessing"><a href="#scrna-seq-preprocessing" class="header-anchor">#</a> scRNA-seq preprocessing</h2> <h3 id="indexing"><a href="#indexing" class="header-anchor">#</a> Indexing</h3> <p>Before read alignment and quality control, a one-time genome indexing
needs to be done. The command line using <code>STAR</code> will look like this:</p> <div class="language- extra-class"><pre class="language-text"><code>$ STAR --runMode genomeGenerate --runThreadN 4 --genomeDir STARindex 
--genomeFastaFiles chr21.fa --sjdbGTFfile Homo_sapiens.GRCh37.75.21.gtf
</code></pre></div><p>We can equivalently index the genome using the <code>Rcwl</code> tool of
<code>STARindex</code> within <em>R</em>, which was internally passed as cwl scripts, by
only assigning values to the input parameters, and execute the cwl
script using one of the execution functions, e.g., <code>runCWL</code> in the
local computer. Then the output files are ready to pass as input to
the next tool for single cell read alignment.</p> <p>Assign values to the input parameters:</p> <div class="language-r extra-class"><pre class="language-r"><code>STARindex<span class="token operator">$</span>genomeFastaFiles <span class="token operator">&lt;-</span> file.path<span class="token punctuation">(</span>path<span class="token punctuation">,</span> <span class="token string">&quot;chr21.fa&quot;</span><span class="token punctuation">)</span>
STARindex<span class="token operator">$</span>sjdbGTFfile <span class="token operator">&lt;-</span> file.path<span class="token punctuation">(</span>path<span class="token punctuation">,</span> <span class="token string">&quot;Homo_sapiens.GRCh37.75.21.gtf&quot;</span><span class="token punctuation">)</span>
runCWL<span class="token punctuation">(</span>cwl <span class="token operator">=</span> STARindex<span class="token punctuation">,</span> outdir <span class="token operator">=</span> file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARindex_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">,</span> docker <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
dir<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARindex_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">,</span> recursive <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>  <span class="token comment">## output files</span>
</code></pre></div><p>Note that the <code>docker</code> argument in <code>runCWL</code> function takes 4 values:</p> <ol><li>TRUE (default, recommended), which automatically pulls docker
images for the required command line tools.</li> <li>FALSE, if users have already pre-installed all required command
line tools.</li> <li>&quot;singularity&quot; if the running environment doesn't support docker but
singularity.</li> <li>&quot;udocker&quot; for the docker-like runtime without any administrator
privileges.</li></ol> <h3 id="alignment"><a href="#alignment" class="header-anchor">#</a> Alignment</h3> <p>Assign values to input parameters:</p> <div class="language-r extra-class"><pre class="language-r"><code>cdna.fastq <span class="token operator">&lt;-</span> file.path<span class="token punctuation">(</span>path<span class="token punctuation">,</span> list.files<span class="token punctuation">(</span>path<span class="token punctuation">,</span> pattern <span class="token operator">=</span> <span class="token string">&quot;_R2_&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
cb.fastq <span class="token operator">&lt;-</span> file.path<span class="token punctuation">(</span>path<span class="token punctuation">,</span> list.files<span class="token punctuation">(</span>path<span class="token punctuation">,</span> pattern <span class="token operator">=</span> <span class="token string">&quot;_R1_&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
cblist <span class="token operator">&lt;-</span> file.path<span class="token punctuation">(</span>path<span class="token punctuation">,</span> <span class="token string">&quot;subset15_demo_barcode.txt&quot;</span><span class="token punctuation">)</span>
genomeDir <span class="token operator">&lt;-</span> file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARindex_output/STARindex&quot;</span><span class="token punctuation">)</span>

inputs<span class="token punctuation">(</span>STARsolo<span class="token punctuation">)</span>
STARsolo<span class="token operator">$</span>readFilesIn_cdna <span class="token operator">&lt;-</span> cdna.fastq
STARsolo<span class="token operator">$</span>readFilesIn_cb <span class="token operator">&lt;-</span> cb.fastq
STARsolo<span class="token operator">$</span>whiteList <span class="token operator">&lt;-</span> cblist
STARsolo<span class="token operator">$</span>genomeDir <span class="token operator">&lt;-</span> genomeDir

runCWL<span class="token punctuation">(</span>STARsolo<span class="token punctuation">,</span> outdir <span class="token operator">=</span> file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARsolo_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
dir<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARsolo_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">,</span> recursive <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
</code></pre></div><p>The output files generated in the &quot;STARsolo_output&quot; folder can now be
passed into the next tool for QC.</p> <h3 id="count-filtering"><a href="#count-filtering" class="header-anchor">#</a> Count filtering</h3> <p>To get a high-quality count matrix we apply the <code>DropletUtils</code> <em>Bioconductor</em> package, which will produce a filtered dataset that is
more representative of the <code>Cell Ranger</code> pipeline.</p> <p>Since CWL itself doesn't support the integration of <em>R</em> packages or
<em>R</em> function, this is a unique feature for <code>Rcwl</code>, where we can easily
connect the upstream data preprocessing steps (previously based on
command line tools) and the downstream data analysis steps (heavily
done in <em>R/Bioconductor</em>).</p> <p>The idea here is to put anything you need into a user-defined <em>R</em>
function, with specified arguments for input and output files, then
it's ready to be wrapped as an <code>Rcwl</code> tools for execution.</p> <p>For example, in wrapping the <em>Bioconductor</em> package
<a href="https://bioconductor.org/packages/DropletUtils/" target="_blank" rel="noopener noreferrer">DropletUtils<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>
functionalities, we wrote this <code>Rcwl</code> tool called
<a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/tl_DropletUtils.R" target="_blank" rel="noopener noreferrer">tl_DropletUtils<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>
with 3 major steps: 1) use the <code>read10xCounts</code> function to read the
raw aligned files and convert into a <code>SingleCellExperiment</code> object. 2)
calculate the barcode ranks and plotting. 3) calculate the empty
droplets and plotting. We have also defined the tool output to collect
the <code>SingleCellExperiment</code> Rdata file and diagnostic figures pdf file.</p> <div class="language-r extra-class"><pre class="language-r"><code>library<span class="token punctuation">(</span>DropletUtils<span class="token punctuation">)</span>
inputs<span class="token punctuation">(</span>DropletUtils<span class="token punctuation">)</span>
DropletUtils<span class="token operator">$</span>dirname <span class="token operator">&lt;-</span> file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARsolo_output/Solo.out&quot;</span><span class="token punctuation">)</span>
DropletUtils<span class="token operator">$</span>lower <span class="token operator">&lt;-</span> <span class="token number">100</span>
DropletUtils<span class="token operator">$</span>df <span class="token operator">&lt;-</span> <span class="token number">5</span>
runCWL<span class="token punctuation">(</span>DropletUtils<span class="token punctuation">,</span> outdir <span class="token operator">=</span> file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;dropletUtils_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">,</span> showLog <span class="token operator">=</span> <span class="token boolean">FALSE</span><span class="token punctuation">)</span>
dir<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;dropletUtils_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre></div><p>Now that we get 2 output files:</p> <ol><li>The pdf file with 2 diagnostic figures: Barcode ranks, and empty
droplets. For details regarding interpretation of each diagnostic
figure, please refer to the <a href="https://bioconductor.org/packages/release/bioc/vignettes/DropletUtils/inst/doc/DropletUtils.html" target="_blank" rel="noopener noreferrer">DropletUtils
vignette<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>.</li></ol> <img src="/case_studies/scrnaseq/diag1.png" alt="barcode ranks"> <img src="/case_studies/scrnaseq/diag2.png" alt="empty droplets"> <div class="language-r extra-class"><pre class="language-r"><code>Biobase<span class="token operator">::</span>openPDF<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;dropletUtils_output/diagnostics.pdf&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre></div><ol start="2"><li>The <code>SingleCellExperiment</code> object which has filtered out
unqualified cells and is analysis ready.</li></ol> <div class="language-r extra-class"><pre class="language-r"><code>sce <span class="token operator">&lt;-</span> readRDS<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;dropletUtils_output/sce_filtered.rds&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
sce
</code></pre></div><h3 id="scrna-seq-preprocessing-pipeline"><a href="#scrna-seq-preprocessing-pipeline" class="header-anchor">#</a> scRNA-seq preprocessing pipeline</h3> <p>Alternatively and more easily, we can use the pipeline called
<code>pl_STARsoloDropletUtils</code> (recipe
<a href="https://github.com/rworkflow/RcwlRecipes/blob/master/Rcwl/pl_STARsoloDropletUtils.R" target="_blank" rel="noopener noreferrer">here<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a>)
for the scRNA-seq data preprocessing. This pipeline integrated the
<code>STARsolo</code> and <code>DropletUtils</code> for a streamlined preprocessing analysis
within <em>R</em>. It is included in <code>RcwlPipelines</code> and is ready to be
customized for your own research. Pipelines can be visualized by the
<code>plotCWL</code> function.</p> <div class="language-r extra-class"><pre class="language-r"><code>STARsoloDropletUtils <span class="token operator">&lt;-</span> cwlLoad<span class="token punctuation">(</span><span class="token string">&quot;pl_STARsoloDropletUtils&quot;</span><span class="token punctuation">)</span>
</code></pre></div><div class="language-r extra-class"><pre class="language-r"><code>plotCWL<span class="token punctuation">(</span>STARsoloDropletUtils<span class="token punctuation">)</span>
</code></pre></div><img src="/case_studies/scrnaseq/workflow.png" alt="scRNA-seq preprocessing workflow"> <p>For <code>Rcwl</code> pipelines, we only need to assign input values for the
whole pipeline, not individual tools involved. The input and output
between each step are pre-defined in the pipeline to ensure a smooth
passing.</p> <div class="language-r extra-class"><pre class="language-r"><code>inputs<span class="token punctuation">(</span>STARsoloDropletUtils<span class="token punctuation">)</span>
STARsoloDropletUtils<span class="token operator">$</span>fastq_cdna <span class="token operator">&lt;-</span> cdna.fastq
STARsoloDropletUtils<span class="token operator">$</span>fastq_cb <span class="token operator">&lt;-</span> cb.fastq
STARsoloDropletUtils<span class="token operator">$</span>genomeDir <span class="token operator">&lt;-</span> file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARindex_output/STARindex&quot;</span><span class="token punctuation">)</span>
STARsoloDropletUtils<span class="token operator">$</span>whiteList <span class="token operator">&lt;-</span> cblist
STARsoloDropletUtils<span class="token operator">$</span>runThreadN <span class="token operator">&lt;-</span> <span class="token number">1</span>

runCWL<span class="token punctuation">(</span>STARsoloDropletUtils<span class="token punctuation">,</span> outdir <span class="token operator">=</span> file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;scpipeline_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre></div><p>The overall output of the pipeline was pre-defined to glob the
important files from separate steps.</p> <div class="language-r extra-class"><pre class="language-r"><code>outputs<span class="token punctuation">(</span>STARsoloDropletUtils<span class="token punctuation">)</span>
dir<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;scpipeline_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">,</span> recursive <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>
</code></pre></div><h3 id="submit-parallel-jobs-for-starsolo"><a href="#submit-parallel-jobs-for-starsolo" class="header-anchor">#</a> Submit parallel jobs for <code>STARsolo</code></h3> <p>Powered by <code>BiocParallel</code>,<code>Rcwl</code> supports parallel job running for
multiple samples using the <code>runCWLBatch</code> function. The following
example demonstrates how to do the parallel alignment for the 2
samples using the <code>STARsolo</code> tool.</p> <p>The <code>BPPARAM</code> argument in <code>runCWLBatch()</code> defines the parallel
parameters. It can be defined by <code>BiocParallel::BatchtoolsParam</code>
function, where the <code>cluster</code> argument takes different values for
different cluster job manager, such as &quot;multicore&quot;, &quot;sge&quot; and
&quot;slurm&quot;. More details about available options can be checked by
<code>?BiocParallel::BatchtoolsParam</code>.</p> <div class="language-r extra-class"><pre class="language-r"><code>library<span class="token punctuation">(</span>BiocParallel<span class="token punctuation">)</span>
bpparam <span class="token operator">&lt;-</span> BatchtoolsParam<span class="token punctuation">(</span>workers <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">,</span> cluster <span class="token operator">=</span> <span class="token string">&quot;sge&quot;</span><span class="token punctuation">,</span>
                           template <span class="token operator">=</span> batchtoolsTemplate<span class="token punctuation">(</span><span class="token string">&quot;sge&quot;</span><span class="token punctuation">)</span><span class="token punctuation">)</span>
</code></pre></div><p>In the following example, we are using &quot;multicore&quot; for the parallel
running.</p> <p>The <code>inputList</code> argument is required to be a list of input parameter
values for samples that are to be computed parallelly. <strong>NOTE</strong> that
the names of the list must be consistent with the ids of input
parameters. In this example, the names are <code>readFilesIn_cdna</code> and
<code>readFilesIn_cb</code>.</p> <p>The <code>paramList</code> argument is required to be a list of input parameter
values that are to be shared for all parallelly running samples.</p> <div class="language-r extra-class"><pre class="language-r"><code>bpparam <span class="token operator">&lt;-</span> BatchtoolsParam<span class="token punctuation">(</span>workers <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">,</span> cluster <span class="token operator">=</span> <span class="token string">&quot;multicore&quot;</span><span class="token punctuation">)</span>
input_lst <span class="token operator">&lt;-</span> list<span class="token punctuation">(</span>readFilesIn_cdna <span class="token operator">=</span> list<span class="token punctuation">(</span>
                      sample1 <span class="token operator">=</span> cdna.fastq<span class="token punctuation">,</span>
                      sample2 <span class="token operator">=</span> cdna.fastq<span class="token punctuation">)</span><span class="token punctuation">,</span>
                  readFilesIn_cb <span class="token operator">=</span> list<span class="token punctuation">(</span>
                      sample1 <span class="token operator">=</span> cb.fastq<span class="token punctuation">,</span>
                      sample2 <span class="token operator">=</span> cb.fastq<span class="token punctuation">)</span>
                  <span class="token punctuation">)</span>
param_lst <span class="token operator">&lt;-</span> list<span class="token punctuation">(</span>whiteList <span class="token operator">=</span> cblist<span class="token punctuation">,</span>
                  genomeDir <span class="token operator">=</span> genomeDir<span class="token punctuation">,</span>
                  runThreadN <span class="token operator">=</span> <span class="token number">2</span><span class="token punctuation">)</span>
res <span class="token operator">&lt;-</span> runCWLBatch<span class="token punctuation">(</span>cwl <span class="token operator">=</span> STARsolo<span class="token punctuation">,</span>
                   outdir <span class="token operator">=</span> file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARsolo_batch_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">,</span>
                   inputList <span class="token operator">=</span> input_lst<span class="token punctuation">,</span> paramList <span class="token operator">=</span> param_lst<span class="token punctuation">,</span> 
                   BPPARAM <span class="token operator">=</span> bpparam<span class="token punctuation">)</span>
</code></pre></div><p>The results are saved in separate folders for each parallel sample.</p> <div class="language-r extra-class"><pre class="language-r"><code>dir<span class="token punctuation">(</span>file.path<span class="token punctuation">(</span>outpath<span class="token punctuation">,</span> <span class="token string">&quot;STARsolo_batch_output&quot;</span><span class="token punctuation">)</span><span class="token punctuation">,</span> recursive <span class="token operator">=</span> <span class="token boolean">TRUE</span><span class="token punctuation">)</span>  <span class="token comment">## output files</span>
</code></pre></div><h3 id="shiny-interface"><a href="#shiny-interface" class="header-anchor">#</a> Shiny interface</h3> <p><code>cwlShiny()</code> opens a user-friendly shiny interface for running any
<code>Rcwl</code> tools or pipelines. By default, users need to put in the
absolute file path for each input parameter. <strong>NOTE</strong>, multiple file
paths need to be separated by colon. Click the <code>run</code> button, it will
start running and return the output file paths under <code>Output</code>
tag. Users can also check the <code>Command</code> and <code>Log</code> in the shiny
interface page.</p> <div class="language-r extra-class"><pre class="language-r"><code>cwlShiny<span class="token punctuation">(</span>STARsolo<span class="token punctuation">)</span>
</code></pre></div><img src="/case_studies/scrnaseq/cwlShiny_STARsolo.png" alt="cwlShiny STARsolo"> <h2 id="sessioninfo"><a href="#sessioninfo" class="header-anchor">#</a> SessionInfo</h2> <div class="language-r extra-class"><pre class="language-r"><code>sessionInfo<span class="token punctuation">(</span><span class="token punctuation">)</span>
</code></pre></div><h2 id="links"><a href="#links" class="header-anchor">#</a> Links</h2> <ol><li><a href="https://github.com/rworkflow/Rcwl_scRNAseq" target="_blank" rel="noopener noreferrer">GitHub repository<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> for the workshop materials and instructions.</li> <li><a href="https://youtu.be/5jkJAfkjFGo" target="_blank" rel="noopener noreferrer">YouTube link<span><svg xmlns="http://www.w3.org/2000/svg" aria-hidden="true" focusable="false" x="0px" y="0px" viewBox="0 0 100 100" width="15" height="15" class="icon outbound"><path fill="currentColor" d="M18.8,85.1h56l0,0c2.2,0,4-1.8,4-4v-32h-8v28h-48v-48h28v-8h-32l0,0c-2.2,0-4,1.8-4,4v56C14.8,83.3,16.6,85.1,18.8,85.1z"></path> <polygon fill="currentColor" points="45.7,48.7 51.3,54.3 77.2,28.5 77.2,37.2 85.2,37.2 85.2,14.9 62.8,14.9 62.8,22.9 71.5,22.9"></polygon></svg> <span class="sr-only">(opens new window)</span></span></a> for the Rcwl workshop on BioCAsia2020.</li></ol></div> <footer class="page-edit"><!----> <!----></footer> <!----> </main></div><div class="global-ui"></div></div>
    <script src="/assets/js/app.78d66748.js" defer></script><script src="/assets/js/2.77543436.js" defer></script><script src="/assets/js/61.31d1b0d6.js" defer></script>
  </body>
</html>
