(window.webpackJsonp=window.webpackJsonp||[]).push([[65],{385:function(t,e,a){"use strict";a.r(e);var i=a(43),l=Object(i.a)({},(function(){var t=this,e=t.$createElement,a=t._self._c||e;return a("ContentSlotsDistributor",{attrs:{"slot-key":t.$parent.slotKey}},[a("h2",{attrs:{id:"bowtie2-index"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#bowtie2-index"}},[t._v("#")]),t._v(" bowtie2_index")]),t._v(" "),a("p",[t._v("Bowtie 2 is an ultrafast and memory-efficient tool for aligning sequencing reads to long reference sequences. "),a("code",[t._v("bowtie2-build")]),t._v(" is used to build the index files."),a("br"),t._v("Recipe source code: "),a("a",{attrs:{href:"https://github.com/rworkflow/ReUseDataRecipe/blob/master/bowtie2_index.R",target:"_blank",rel:"noopener noreferrer"}},[t._v("https://github.com/rworkflow/ReUseDataRecipe/blob/master/bowtie2_index.R"),a("OutboundLink")],1),a("br"),t._v("Data source: "),a("a",{attrs:{href:"https://bowtie-bio.sourceforge.net/bowtie2/index.shtml",target:"_blank",rel:"noopener noreferrer"}},[t._v("https://bowtie-bio.sourceforge.net/bowtie2/index.shtml"),a("OutboundLink")],1)]),t._v(" "),a("h2",{attrs:{id:"plot"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#plot"}},[t._v("#")]),t._v(" plot")]),t._v(" "),a("p",[a("img",{attrs:{src:"/plots/bowtie2_index.svg",alt:"## bowtie2_index"}})]),t._v(" "),a("h2",{attrs:{id:"inputs"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#inputs"}},[t._v("#")]),t._v(" Inputs")]),t._v(" "),a("table",[a("thead",[a("tr",[a("th",{staticStyle:{"text-align":"left"}}),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("label")]),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("type")]),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("description")])])]),t._v(" "),a("tbody",[a("tr",[a("td",{staticStyle:{"text-align":"left"}},[t._v("genome")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("reference genome")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("File")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("The reference genome file")])])])]),t._v(" "),a("h2",{attrs:{id:"outputs"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#outputs"}},[t._v("#")]),t._v(" Outputs")]),t._v(" "),a("table",[a("thead",[a("tr",[a("th",{staticStyle:{"text-align":"left"}}),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("label")]),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("type")]),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("description")])])]),t._v(" "),a("tbody",[a("tr",[a("td",{staticStyle:{"text-align":"left"}},[t._v("btIdx")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("bowtie2 index")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("File[]")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("A list of bowtie index files")])])])]),t._v(" "),a("h2",{attrs:{id:"example"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#example"}},[t._v("#")]),t._v(" Example:")]),t._v(" "),a("div",{staticClass:"language- extra-class"},[a("pre",{pre:!0,attrs:{class:"language-text"}},[a("code",[t._v("bowtie2_index <- recipeLoad('bowtie2_index')\nbowtie2_index$genome <- 'GRCh38.primary_assembly.genome.fa' ## need to be a valid file path\ngetData(bowtie2_index, outdir = 'data/folder', notes = c('bowtie2_index', 'GRCh38.primary_assembly'))\n\n## Get data from local catch\ndataUpdate('data/folder')\ndataSearch(c('bowtie2_index', 'GRCh38'))\n\n")])])])])}),[],!1,null,null,null);e.default=l.exports}}]);