(window.webpackJsonp=window.webpackJsonp||[]).push([[77],{374:function(t,e,a){"use strict";a.r(e);var l=a(43),s=Object(l.a)({},(function(){var t=this,e=t.$createElement,a=t._self._c||e;return a("ContentSlotsDistributor",{attrs:{"slot-key":t.$parent.slotKey}},[a("h2",{attrs:{id:"salmon-index"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#salmon-index"}},[t._v("#")]),t._v(" salmon_index")]),t._v(" "),a("p",[t._v("Salmon is a tool for wicked-fast transcript quantification from RNA-seq data. The recipe is to build a salmon index for your transcriptome. The script is prepared according to this instruction (https://combine-lab.github.io/alevin-tutorial/2019/selective-alignment/.)"),a("br"),t._v("Recipe source code: "),a("a",{attrs:{href:"https://github.com/rworkflow/ReUseDataRecipe/blob/master/salmon_index.R",target:"_blank",rel:"noopener noreferrer"}},[t._v("https://github.com/rworkflow/ReUseDataRecipe/blob/master/salmon_index.R"),a("OutboundLink")],1),a("br"),t._v("Data source: "),a("a",{attrs:{href:"https://salmon.readthedocs.io/en/latest/salmon.html",target:"_blank",rel:"noopener noreferrer"}},[t._v("https://salmon.readthedocs.io/en/latest/salmon.html"),a("OutboundLink")],1)]),t._v(" "),a("h2",{attrs:{id:"plot"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#plot"}},[t._v("#")]),t._v(" plot")]),t._v(" "),a("p",[a("img",{attrs:{src:"/plots/salmon_index.svg",alt:"## salmon_index"}})]),t._v(" "),a("h2",{attrs:{id:"inputs"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#inputs"}},[t._v("#")]),t._v(" Inputs")]),t._v(" "),a("table",[a("thead",[a("tr",[a("th",{staticStyle:{"text-align":"left"}}),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("label")]),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("type")]),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("description")])])]),t._v(" "),a("tbody",[a("tr",[a("td",{staticStyle:{"text-align":"left"}},[t._v("genome")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("reference genome")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("File")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("The fasta file for reference genome")])]),t._v(" "),a("tr",[a("td",{staticStyle:{"text-align":"left"}},[t._v("transcript")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("transcripts")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("File")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("The fasta file for transcripts from gencode")])])])]),t._v(" "),a("h2",{attrs:{id:"outputs"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#outputs"}},[t._v("#")]),t._v(" Outputs")]),t._v(" "),a("table",[a("thead",[a("tr",[a("th",{staticStyle:{"text-align":"left"}}),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("label")]),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("type")]),t._v(" "),a("th",{staticStyle:{"text-align":"left"}},[t._v("description")])])]),t._v(" "),a("tbody",[a("tr",[a("td",{staticStyle:{"text-align":"left"}},[t._v("salmonIdx")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("salmon index")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("Directory")]),t._v(" "),a("td",{staticStyle:{"text-align":"left"}},[t._v("A directory containing salmon index files")])])])]),t._v(" "),a("h2",{attrs:{id:"example"}},[a("a",{staticClass:"header-anchor",attrs:{href:"#example"}},[t._v("#")]),t._v(" Example:")]),t._v(" "),a("div",{staticClass:"language- extra-class"},[a("pre",{pre:!0,attrs:{class:"language-text"}},[a("code",[t._v("salmon_index <- recipeLoad('salmon_index.R')\nsalmon_index$genome <- 'GRCh38.primary_assembly.genome.fa'\nsalmon_index$transcript <- 'gencode.v42.transcripts.fa'\ngetData(salmon_index, outdir = 'data/folder', notes = c('salmon_index', 'GRCh38.primary_assembly', 'gencode.v42'))\n\n## Get data from local catch\ndataUpdate('data/folder')\ndataSearch(c('salmon_index', 'GRCh38.primary_assembly', 'gencode.v42'))\n\n")])])])])}),[],!1,null,null,null);e.default=s.exports}}]);